### PR TITLE
Implement Cutout in Numpy, PyTorch, and TensorFlow

### PR DESCRIPTION
**Pull request by f4str**
# Description

Implementation of the Cutout data augmentation defense in the Numpy, PyTorch, and TensorFlow frameworks.

Fixes # (issue)

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Tests for the Cutout preprocessor in Numpy
- [x] Tests for the Cutout preprocessor in PyTorch
- [x] Tests for the Cutout preprocessor in TensorFlow

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by f4str** [comment]
@beat-buesser once you've taken a look, please let me know if you think there should really be a different implementation for each framework. I don't believe there is any benefit from having framework specific implementations for PyTorch and TensorFlow since there is not much being done that offers a speedup from GPU. I think only having one Numpy version would make more sense here, but I'm open to suggestions.

**Comment by beat-buesser** [comment]
Hi @f4str That's a good question. Can we think of an application where accurate gradient back-propagation would be useful? E.g for an adaptive poisoning attack on DP-InstaHide?

**Comment by f4str** [comment]
> Hi @f4str That's a good question. Can we think of an application where accurate gradient back-propagation would be useful? E.g for an adaptive poisoning attack on DP-InstaHide?

Hi @beat-buesser thank you for the response. There may be some scenarios where accurate gradient backprop is useful. Adaptive poisoning attacks are definitely one use-case since the DP-InstaHide paper does actually evaluate against adaptive attacks, specifically using gradient matching (Witches' Brew). Whitebox evasion attacks like PGD might also be a use-case since that requires full gradients to the original image.

**Comment by beat-buesser** [comment]
Hi @f4str The augmentations are random, but an adaptive still needs accurate gradients corresponding to the respective randomly sampled augmentation.

**Comment by f4str** [comment]
> Hi @f4str The augmentations are random, but an adaptive still needs accurate gradients corresponding to the respective randomly sampled augmentation.

Hi @beat-buesser that is true. In that case, it does make sense to have framework specific implementations to ensure that gradients are accurate. This is already the case, I'll just continue as is.

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for contributing the first set of data augmentation preprocessors and congratulations to your first contribution to ART!
